text/0000-node-reactions.md

```diff
@@ -0,0 +1,208 @@
+
+*Proposal 1*: One simple API that accepts hooks for various lifecycle events
+```js
+/**
+ * @param {Node} node - Node that should be observed
+ * @param {String} reactionEventType - one of connected, disconnected, adopted etc
+ * @param {Function} callback
+ */
+function reactTo(node, reactionEventType, callback);
+```
+
+*Proposal 2*: One API per lifecycle event
+```js
+addConnectedCallback(node, callback);
+addDisconnectedCallback(node, callback);
+```
+
+# Motivation
+The browser provides ways to react to lifecycle events in the window or document. For example, [`DOMContentLoaded`](https://developer.mozilla.org/en-US/docs/Web/API/Window/DOMContentLoaded_event) event signals that the initial HTML document has been loaded and parsed. The [`load`](https://developer.mozilla.org/en-US/docs/Web/API/Window/load_event) event signals that the page is full loaded. Similarly, [`beforeunload`](https://developer.mozilla.org/en-US/docs/Web/API/WindowEventHandlers/onbeforeunload) allows reactions to the resources being unloaded.
+
+Web Components supports reacting to the lifecycle of a custom element. Hooks for  lifecycle events are part of the custom component definition. Read the spec [here](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_custom_elements#Using_the_lifecycle_callbacks) for more detail. 
+
+Lightning Web Components(LWC) has support for similar lifecycle events. Of particular interest to this RFC is the `connectedCallback` and `disconnectedCallback`. In LWC, the responsibility of invoking the hooks is distributed between the engine itself and snabdom. The goal of this RFC is to abstract the responsibility of managing the lifecycle into a library and make LWC subscribe to lifecycle events from the node-reactions library.
+
+Further, there are inconsistencies in the current implementation. For example, connectedCallback should only be invoked if a custom element is connected to the document. In LWC, connectedCallback is invoked anytime a custom element is connected to a parent, regardless of wether the parent is connected to the document or not.
+
+# Detailed design
+
+## When does a node qualify as connected?
+
+For custom elements, a node that is appended to the document or one of its connected discendents is considered as connected. [Node.isConnected](https://dom.spec.whatwg.org/#dom-node-isconnected) offers the most reliable way to determine if a node is connected.
+
+## Invariants
+1. A node can be connected and disconnected from the DOM more than once. Every time the event happens, the qualifying hook should be invoked.
+2. For a given node, more than one service can be registered to react to the same lifecycle event. In which case, each hook should be invoked in the order of registration(FIFO). 
+    2.1 Duplicate hooks for the same event type, for the same node will be ignored. For example, if a service registers the same connectedCallback more than once for the same node, the callback is invoked only once.
```

Comment:
is there a particular reason for this? seems to slow down things... while event listeners and other similar mechanisms do not apply this filtering.

Reply:
Duplicate event listeners on the same node do get filtered out. Here's an example to demonstrate: https://jsbin.com/weyewul/edit?html,js,console,output